---
title: "Pre-process the data for PENCIL input"
author: "Tiangen Chang tiangen.chang@nih.gov"
output: 
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 2
vignette: >
  %\VignetteIndexEntry{}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r}
# clear all objects including hidden objects
rm(list = ls(all.names = TRUE)) 
# free up memory and report the memory usage
gc()
```


```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      comment = FALSE, 
                      warning = FALSE, 
                      message = FALSE)
require(ggplot2)
ggplot2::theme_set(new = theme_bw(base_size = 12)) # sets the theme and axis font size
```



# Data description:

1. GSE200996

Original paper: Luoma, Adrienne M., et al. "Tissue-resident memory and circulating T cells are early responders to pre-surgical cancer immunotherapy." Cell 185.16 (2022): 2918-2935. https://doi.org/10.1016/j.cell.2022.06.018

Data type: matched tumor and blood, scRNA-seq, scTCR-seq, bulk TCR; Samples: from 30 patients from 2 different cohorts (anti-PD-1 monotherapy, anti-PD-1+anti-CTLA-4 combination therapy)


# Data input:

Data matrix downloaded from GEO

# Sample description (no healthy donors)

Blood B1 (pre): P01, P02, P04, P05, P08, P09, P10, P12, P13, P14, P15, P16, P17, P18, P19, P20, P21, P22, P23, P24, P25, P26, P28, P29, P30, P31, P32

Blood B2 (post): P01, P02, P04, P05, P08, P09, P10, P12, P13, P14, P15, P16, P17, P19, P20, P22, P23, P24, P26, P28, P29, P30, P31, P32

Blood B3 (follow-up): P01, P02, P04, P05, P09, P10, P12, P13, P14, P15, P17, P19, P20, P21, P22, P23, P25, P32

Tumor pre: P18, P23, P24, P29, P32, [P27 (do not have blood match)]

Tumor post: P13, P14, P15, P16, P18, P19, P20, P21, P22, P23, P24, P25, P26, P27, P28, P29, P30, P31, P32



# Aims:

- integrate all patients' data object

- analysis the integrated data

- annotate the data with celltypist

- save the integrated seurat object

# Connect to server
```{bash}
ssh changt7@biowulf.nih.gov
sinteractive --mem=64g --cpus-per-task=32 --time=24:00:00
sinteractive --mem=128g --cpus-per-task=32 --gres=gpu:k80:2 --time=24:00:00
# --gres=lscratch:1024
module load R/
R

```

# load required package
```{r, include=FALSE}

library(Seurat)
library(GEOquery)
library(dplyr)
library(ggplot2)
library(SingleR)
library(sctransform)
library(celldex)
library(dittoSeq)
library(SeuratWrappers)
library(harmony)
library(edgeR)
library(scuttle)
library(scran)
library(scFeatureFilter)
library(cowplot)
library(patchwork)
library(edgeR)
library(RColorBrewer)
library(stringr)

```

# set input and output directories & parameters
```{r}

dataset = 'GSE120575' # 'GSE200996'   'GSE120575'   'GSE169246'

# local
data_dir <- paste0("/Users/changt7/Documents/00.PostDocWork/CancerResearch/01.Projects/00.Collaborating_projects/05.PENCIL/02.Input/", dataset, '/')
processed_dir <- paste0("../03.Results/", dataset, '/')
fig_dir <- paste0("../03.Results/", dataset, '/Figures/')

# server
shared_data_dir = paste0("/data/Lab_ruppin/tiangen/CancerProject/02.PENCIL/02.Input/")
data_dir <- paste0(shared_data_dir, dataset, '/')
processed_dir <- paste0("/data/Lab_ruppin/tiangen/CancerProject/02.PENCIL/03.Results/", dataset, '/')
fig_dir = paste0("/data/Lab_ruppin/tiangen/CancerProject/02.PENCIL/03.Results/", dataset, "/Figures/")
PENCIL_result_dir = paste0("/data/Lab_ruppin/tiangen/CancerProject/02.PENCIL/01.Scripts/results/")
```



# load seurat object
```{r}

sampleType_name = 'Tissue' # 'PBMC'  'Tissue'

merged_seu = readRDS(file = paste0(data_dir,"seu_",sampleType_name,".rds"))

```


# Add phenotype info for GSE200996 (Luoma et al., HNSCC, ICB dataset)
```{r}

# Read in phenotype info.
phenotype_df = read.csv(file = paste0(data_dir,"clinical_info.csv"))
phenotype_df = phenotype_df[,c(1,7,9,10,15:26)]
colnames(phenotype_df) = c(colnames(phenotype_df)[1:4],'Downstaging','RECIST_response','Volumetric_response','Viable_Tumor', 'Pathological_response', 'Pathological_response_binned', 'Max_response', 'Any_response', 'OS_days', 'PFS_days', 'PFS_event', 'OS_event')

# Add phenotype info to seurat object
meta <- merged_seu@meta.data
meta$cell_names <- rownames(meta) # save row names before merge with phenotype data
meta.new <- merge(meta, phenotype_df, by.x = "patient", by.y = "patient_id")
rownames(meta.new) <- meta.new$cell_names # retrieve rownames 
merged_seu <- AddMetaData(merged_seu, meta.new)
#cell.names <- sapply(seq_along(rownames(merged_seu@meta.data)), function(i) paste0("cell_", i), USE.NAMES = F)
#merged_seu@meta.data$cellID <- cell.names # rename cells to cell_1, cell_2, ...
#merged_seu <- RenameCells(object = merged_seu, new.names = merged_seu@meta.data$cellID)
#print(merged_seu$cellID[1:10])

saveRDS(object = merged_seu, file = paste0(data_dir, "seu_",sampleType_name,".rds")) 



```


# Add phenotype info for GSE169246 (Zeming Zhang, TNBC, ICB dataset)
```{r}

# Read in phenotype info.
phenotype_df = read.csv(file = paste0(data_dir,"clinical_info.csv"))
phenotype_df = phenotype_df[,c(1,7,9,10,15:26)]
colnames(phenotype_df) = c(colnames(phenotype_df)[1:4],'Downstaging','RECIST_response','Volumetric_response','Viable_Tumor', 'Pathological_response', 'Pathological_response_binned', 'Max_response', 'Any_response', 'OS_days', 'PFS_days', 'PFS_event', 'OS_event')

# Add phenotype info to seurat object
meta <- merged_seu@meta.data
meta$cell_names <- rownames(meta) # save row names before merge with phenotype data
meta.new <- merge(meta, phenotype_df, by.x = "patient", by.y = "patient_id")
rownames(meta.new) <- meta.new$cell_names # retrieve rownames 
merged_seu <- AddMetaData(merged_seu, meta.new)
#cell.names <- sapply(seq_along(rownames(merged_seu@meta.data)), function(i) paste0("cell_", i), USE.NAMES = F)
#merged_seu@meta.data$cellID <- cell.names # rename cells to cell_1, cell_2, ...
#merged_seu <- RenameCells(object = merged_seu, new.names = merged_seu@meta.data$cellID)
#print(merged_seu$cellID[1:10])

saveRDS(object = merged_seu, file = paste0(data_dir, "seu_",sampleType_name,".rds")) 

```

# load raw expression and phenotype data and construct seurat object for GSE120575 (Sade-Feldman et al., melanoma, ICB dataset)
```{r}

library(readr)

# read in log2(TPM+1) values (transcript-per-million reads) for 55,737 transcripts (rows) and 16,291 cells (columns)
data <- read.table(paste0(data_dir,"GSE120575_Sade_Feldman_melanoma_single_cells_TPM_GEO.txt"), header = TRUE, sep = "\t", row.names = 1, stringsAsFactors = FALSE, fill = TRUE) 
new_colnames = c(colnames(data)[2:16292],'X')
colnames(data) = new_colnames
data <- data[, -ncol(data)] # remove last NA column
patient_df <- data.frame(patient = t(data[1,]))
colnames(patient_df) = c('patient')
patient_df$patient <- sub("Post_P17_myeloid_enriched", "Post_P17", patient_df$patient)
patient_df$patient <- sub("Post_P19_myeloid_enriched", "Post_P19", patient_df$patient)
patient_df$patient <- sub("Post_P2_T_enriched", "Post_P2", patient_df$patient)
patient_df$patient <- sub("Post_P4_T_enriched", "Post_P4", patient_df$patient)
patient_df$patient <- sub("Post_P10_T_enriched", "Post_P10", patient_df$patient)
patient_df$patient <- sub("Post_P13_T_enriched", "Post_P13", patient_df$patient)
patient_df$patient <- sub("Post_P8_T_enriched", "Post_P8", patient_df$patient)
patient_df$patient <- sub("Post_P16_T_enriched", "Post_P16", patient_df$patient)
patient_df$patient <- sub("Post_P6_T_enriched", "Post_P6", patient_df$patient)
patient_df$cellID = rownames(patient_df)
data <- data[-1,] # remove patientID row



# Create a Seurat object
merged_seu <- CreateSeuratObject(counts = data)
# Assign the normalized counts to merged_seu[["RNA"]]@data
merged_seu[["RNA"]]@data <- as.matrix(merged_seu[["RNA"]]@counts)
# convert from log2(TPM+1) --> TPM
merged_seu[["RNA"]]@data <- 2^merged_seu[["RNA"]]@data-1 
# Create a new Seurat object with the TPM as counts
merged_seu_new <- CreateSeuratObject(counts = merged_seu[["RNA"]]@data)
merged_seu = merged_seu_new

# Assign project name
merged_seu$project <- "GSE120575_Sade_Feldman_melanoma"

# add phenotype info
phenotype_df_raw <- read.table(paste0(data_dir,"GSE120575_singleCells_metaInfo_clean.txt"), sep = "\t", quote = "", header = TRUE, stringsAsFactors = FALSE)
phenotype_df = phenotype_df_raw[,c(5,6,7)] # "CellID" "PatientID" "ResponseInfo" "DrugType"  
colnames(phenotype_df) = c("patient", "ResponseInfo", "DrugType")
duplicated_rows <- duplicated(phenotype_df[, 1])
phenotype_df <- phenotype_df[!duplicated_rows, ]
# keep only non-duplicated patientID --> dim(phenotype_df.new)=c(48,3)
phenotype_df.new <- merge(patient_df, phenotype_df, by.x = "patient", by.y = "patient")
rownames(phenotype_df.new) <- phenotype_df.new$cellID


# Add phenotype info to seurat object
# merged_seu@meta.data$timePt <- NULL
# merged_seu@meta.data$patient <- NULL
# merged_seu@meta.data$meta <- NULL
# merged_seu@meta.data$ResponseInfo <- NULL
# merged_seu@meta.data$DrugType <- NULL

meta <- merged_seu@meta.data
meta$cell_names <- rownames(meta) # save row names before merge with phenotype data
meta.new <- merge(meta, phenotype_df.new, by.x = "cell_names", by.y = "cellID")
rownames(meta.new) <- meta.new$cell_names # retrieve rownames 
meta.new <- meta.new[, !(names(meta.new) %in% c("cellID", "cell_names"))]
merged_seu <- AddMetaData(merged_seu, meta.new)

merged_seu$meta = merged_seu$patient
merged_seu$timePt <- ifelse(grepl("^Pre_", merged_seu$meta), "pre", "post")
merged_seu$patient <- gsub("^Pre_|^Post_", "", merged_seu$meta)


#################################################################################
############# Cell annotation using 20.Cell_Annotation.Rmd pipeline #############
#################################################################################

##### Optional: Re-do CD8T cell annotation using original annotation by Sade-Feldman et al. 2018 Cell 

CD8T_cellID = read.csv(file = paste0(data_dir,"CD8T_original_cellAnno.csv"))

CD8T_cellID$Cell.Name <- gsub("-", ".", CD8T_cellID$Cell.Name)

merged_seu@meta.data$cell_type_original <- "Other"

merged_seu@meta.data$cell_type_original[rownames(merged_seu@meta.data) %in% CD8T_cellID$Cell.Name] <- "CD8+ T-cells"

saveRDS(object = merged_seu, file = paste0(data_dir, "seu_",sampleType_name,".rds")) 

```


# subset seurat to focus on cell types of interest (e.g. CD8T, CD4T, B, ...)
```{r}

#devtools::install_github("mojaveazure/seurat-disk")
#devtools::install_github('satijalab/seurat-data')
library(SeuratData)
library(SeuratDisk)

phenotype_name = 'ResponseInfo' # GSE200996: 'Any_response'; GSE120575: 'ResponseInfo'
cellType_of_interest = 'CD8T' # 'CD8T'  'CD4T'  'B'


if (1){ #### use original cell type or multi-method manually labeled cell types
  if (cellType_of_interest == 'CD8T'){
    if (dataset == "GSE120575"){ # 'CD8T' cell annotation from the original paper
      merged_seu_subset <- subset(merged_seu, subset = cell_type_original %in% c("CD8+ T-cells"))
    }else{
      merged_seu_subset <- subset(merged_seu, subset = cell_type %in% c("CD8+ T-cells"))
    }
  }else if (cellType_of_interest == 'CD4T'){
    merged_seu_subset <- subset(merged_seu, subset = cell_type %in% c("CD4+ T-cells", "Tregs"))
  }else if (cellType_of_interest == 'B'){
    merged_seu_subset <- subset(merged_seu, subset = cell_type %in% c("Naive B-cells", "Memory B-cells", "Plasma cells"))
  }
}else if (0){ # use celltypist annotated cell types
  if (cellType_of_interest == 'CD8T'){
    merged_seu_subset <- subset(merged_seu, subset = celltypist.IAL %in% c("CD8+ T-cells"))
  }else if (cellType_of_interest == 'CD4T'){
    merged_seu_subset <- subset(merged_seu, subset = celltypist.IAL %in% c("CD4+ T-cells", "Tregs"))
  }else if (cellType_of_interest == 'B'){
    merged_seu_subset <- subset(merged_seu, subset = celltypist.IAL %in% c("Naive B-cells", "Memory B-cells", "Plasma cells", "Other B-cells"))
  }
}else if (0){ # use singler HPCA annotated cell types
  if (cellType_of_interest == 'CD8T'){
    merged_seu_subset <- subset(merged_seu, subset = HPCA.fine %in% c("CD8+ T-cells"))
  }else if (cellType_of_interest == 'CD4T'){
    merged_seu_subset <- subset(merged_seu, subset = HPCA.fine %in% c("CD4+ T-cells", "Tregs"))
  }else if (cellType_of_interest == 'B'){
    merged_seu_subset <- subset(merged_seu, subset = HPCA.fine %in% c("Naive B-cells", "Memory B-cells", "Plasma cells", "Other B-cells"))
  }
}

merged_seu_subset = merged_seu

# filter genes to only focus on protein coding genes (Safe-Feldman paper list)
protein_coding_genes=read.csv(file = paste0(shared_data_dir,"Feldman_protein_coding_genes.csv"), header=T,colClasses = "character")[,1]
genes <- intersect(protein_coding_genes, row.names(merged_seu_subset))
merged_seu_subset <- merged_seu_subset[genes, ]

print(merged_seu_subset)


## raw data is already log2TPM+1, do not need to normalize any more
# merged_seu_subset <- NormalizeData(object=merged_seu_subset, normalization.method="LogNormalize")
merged_seu_subset <- FindVariableFeatures(object=merged_seu_subset, selection.method='vst', nfeatures=2000)

#Calculate Cell Cycle Score (S-G2M Difference)
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
merged_seu_subset<- CellCycleScoring(merged_seu_subset, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
merged_seu_subset$CC.Difference <- merged_seu_subset$S.Score - merged_seu_subset$G2M.Score

# Calculate mitochondrial gene percentage
merged_seu_subset$percent.mt <- PercentageFeatureSet(merged_seu_subset, pattern = "^MT-")

# Calculate ribosomal gene percentage
merged_seu_subset$percent.ribo <- PercentageFeatureSet(merged_seu_subset, pattern = "^RP[SL]")

# Scale the data
merged_seu_subset <- ScaleData(merged_seu_subset)

# Perform PCA
merged_seu_subset <- RunPCA(merged_seu_subset, pc.genes = merged_seu_subset@var.genes)

# Identify the elbow point
#elbow_plot <- ElbowPlot(merged_seu_subset, ndims = 50)  # Adjust 'ndims' as needed
#plot(elbow_plot)
#elbow_point <- IdentifyElbow(elbow_plot)

# Perform UMAP
merged_seu_subset <- RunUMAP(merged_seu_subset, reduction = "pca", dims = 1:10)

# Find neighbors
merged_seu_subset <- FindNeighbors(merged_seu_subset, reduction = "pca", dims=1:10, k.param=20)

# Find clusters
merged_seu_subset <- FindClusters(merged_seu_subset, resolution = 0.8)

# save to .h5ad for PENCIL to read
SaveH5Seurat(merged_seu_subset, filename = paste0(data_dir, "seu_",sampleType_name,"_", cellType_of_interest, ".h5Seurat"), overwrite = TRUE)
Convert(paste0(data_dir, "seu_",sampleType_name,"_", cellType_of_interest, ".h5Seurat"), dest = "h5ad", assay = "RNA", overwrite = TRUE) # .X: only scaled data for HVGs. .raw.X: all genes 
saveRDS(object = merged_seu_subset, file = paste0(data_dir, "seu_",sampleType_name,"_", cellType_of_interest, ".rds")) 



# save to csv, another format of input for PENCIL
exp_data_scaled = merged_seu_subset@assays[["RNA"]]@scale.data[VariableFeatures(merged_seu_subset),]
write.csv(exp_data_scaled, file=paste0(data_dir,"seu_",sampleType_name,'_',cellType_of_interest,".csv"))

cell_label_df = data.frame(label = merged_seu_subset$ResponseInfo)
write.csv(cell_label_df, file=paste0(data_dir,"cell_label_info.csv"))

emd = merged_seu_subset@reductions[["umap"]]@cell.embeddings
write.csv(emd, file=paste0(data_dir, "seu_",sampleType_name,'_',cellType_of_interest,'_embedding_umap.csv'))

#print(unique(merged_seu$patient))
#print(merged_seu$RECIST_response[1:10])
#print(merged_seu$Any_response[1:10])
#print(unique(merged_seu$timePt))
#print(dim(merged_seu))

```

# For GSE120575, compare self annotation and original annotation of CD8T using UMAP
```{r}

# Plot cell_type_original clusters
DimPlot(merged_seu, group.by = "cell_type_original", label = TRUE, repel = TRUE, raster=FALSE)+NoLegend()+
  ggtitle("")
ggsave(filename = paste0(fig_dir, "UMAP_cellTypeOriginalPaper_CD8T_",sampleType_name,".png"), height = 120, width = 140, dpi = 800, units = "mm")

# Plot cell_type CD8T clusters
merged_seu$highlighted.clusters <- ifelse(merged_seu$cell_type %in% c("CD8+ T-cells"), as.character(merged_seu$cell_type), "Other")
DimPlot(merged_seu, group.by = "highlighted.clusters", label = TRUE, repel = TRUE, raster=FALSE)+NoLegend()+
  ggtitle("")
ggsave(filename = paste0(fig_dir, "UMAP_cellTypeSelf_CD8T_",sampleType_name,".png"), height = 120, width = 140, dpi = 800, units = "mm")
```


# Run PENCIL
```{bash}

cd /data/Lab_ruppin/tiangen/CancerProject/02.PENCIL/01.Scripts

python 02.Run_PENCIL.py /data/Lab_ruppin/tiangen/CancerProject/02.PENCIL/02.Input/GSE200996/seu_Tissue_CD8Tcell.h5ad Any_response GSE200996_CD8T multi-classification 1 0


# classification mode
#exp_data = merged_seu@assays[["RNA"]]@scale.data[VariableFeatures(merged_seu),]
# regression mode
#exp_data = as.matrix(merged_seu@assays[["RNA"]]@counts)

```




# Add PENCIL identified phenotypic cell labels 
```{r}



# load seurat object
merged_seu = readRDS(file = paste0(data_dir,"seu_",sampleType_name,'_',cellType_of_interest,".rds"))



predicted_cell_labels = read.csv(paste0(PENCIL_result_dir, dataset, '_', sampleType_name, '_', cellType_of_interest,"/py/", phenotype_name, "/predicted_labels.csv"))
predicted_cell_labels_new <- subset(predicted_cell_labels, select = 'predicted_label')
colnames(predicted_cell_labels_new) = c("PENCIL_pred_label")

# change label names to yes, no, rej
label_names = unique(predicted_cell_labels_new$PENCIL_pred_label)
print(label_names)
predicted_cell_labels_new$PENCIL_pred_label <- gsub("Responder", "yes", predicted_cell_labels_new$PENCIL_pred_label)
predicted_cell_labels_new$PENCIL_pred_label <- gsub("Non-responder", "no", predicted_cell_labels_new$PENCIL_pred_label)
predicted_cell_labels_new$PENCIL_pred_label <- gsub("Rejected", "rej", predicted_cell_labels_new$PENCIL_pred_label)



meta <- merged_seu@meta.data
meta.new <- cbind(meta, predicted_cell_labels_new) # cell ID in PENCIL output does not match the original ones. But it is in the same order.
merged_seu <- AddMetaData(merged_seu, meta.new)
#merged_seu@meta.data$PENCIL_pred_label = NULL

saveRDS(object = merged_seu, file = paste0(data_dir, "seu_",sampleType_name,"_", cellType_of_interest, ".rds")) 


```

# UMAP visualization of patient IDs, R/NR binarized patients, PENCIL predicted cell labels
```{r}

DimPlot(merged_seu, group.by = "seurat_clusters")+
  ggtitle("")
ggsave(filename = paste0(fig_dir, "UMAP_", cellType_of_interest, "_seuratCluster_",sampleType_name,".png"), height = 120, width = 190, dpi = 800, units = "mm")

DimPlot(merged_seu, group.by = "cell_type", label = TRUE, repel = TRUE, raster=FALSE)+NoLegend()+
  ggtitle("")
ggsave(filename = paste0(fig_dir, "UMAP_", cellType_of_interest, "_cell_type_",sampleType_name,".png"), height = 120, width = 190, dpi = 800, units = "mm")


DimPlot(merged_seu, group.by = "seurat_clusters", label = TRUE, repel = TRUE, raster=FALSE)+NoLegend()+

DimPlot(merged_seu, group.by = "meta")+
  ggtitle("")
ggsave(filename = paste0(fig_dir, "UMAP_", cellType_of_interest, "_patientID_",sampleType_name,".png"), height = 120, width = 190, dpi = 800, units = "mm")


DimPlot(merged_seu, group.by = "timePt")+
  ggtitle("")
ggsave(filename = paste0(fig_dir, "UMAP_", cellType_of_interest, "_timePt_",sampleType_name,".png"), height = 120, width = 190, dpi = 800, units = "mm")


DimPlot(merged_seu, group.by = phenotype_name)+
  ggtitle("")
ggsave(filename = paste0(fig_dir, "UMAP_", cellType_of_interest, "_",phenotype_name,'_',sampleType_name,".png"), height = 120, width = 190, dpi = 800, units = "mm")



DimPlot(merged_seu, group.by = "PENCIL_pred_label")+
  ggtitle("")+
  xlab("Cell label")
ggsave(filename = paste0(fig_dir, "UMAP_", cellType_of_interest, "_PENCILpredLabel_",sampleType_name,".png"), height = 120, width = 190, dpi = 800, units = "mm")


```


# Calculate R/NR cell number and ratios in each patient
```{r}

# Calculate the cell number for each patient
grouped <- merged_seu@meta.data %>%
  group_by(meta, PENCIL_pred_label)

# Calculate the number of cells for each patient and predicted_label
cell_counts <- grouped %>%
  summarise(cell_count = n())
cell_counts <- cell_counts[cell_counts$PENCIL_pred_label != "rej", ]

# Create a vector with all unique patient IDs
all_patients <- unique(cell_counts$meta)
print(unique(merged_seu$PENCIL_pred_label))

unique_label = c("yes","no")
# Create a new data frame with all patients having both 'yes' and 'no'
new_df <- data.frame(meta = rep(all_patients, each = 2),
                     PENCIL_pred_label = rep(unique_label, length(all_patients)))
# Merge the new data frame with the original data frame based on patient ID and 'yes_no'
cell_counts_new <- merge(cell_counts, new_df, by = c("meta", "PENCIL_pred_label"), all = TRUE)
# Replace missing values (NA) in the 'values' column with 0
cell_counts_new$cell_count[is.na(cell_counts_new$cell_count)] <- 0


# Calculate the total count for each patient
df_totals <- aggregate(cell_count ~ meta, cell_counts_new, sum)
# Calculate the count ratios for 'yes' and 'no'
df_ratios <- aggregate(cell_count ~ meta + PENCIL_pred_label, cell_counts_new, sum)
df_ratios <- transform(df_ratios, ratio = cell_count / df_totals$cell_count * 100)
df_ratios$meta0 = df_ratios$meta
if (dataset == "GSE200996"){
  df_ratios$meta <- sub("_tumor$", "", df_ratios$meta)
  df_ratios$patient <- str_extract(df_ratios$meta, ".*(?=_)")
}else if (dataset == "GSE120575"){
  df_ratios$patient <- sub("^(.*?)_", "", df_ratios$meta)

}


phenotype_df = merged_seu@meta.data[c('patient', phenotype_name)]
duplicated_rows <- duplicated(phenotype_df[, 1])
phenotype_df <- phenotype_df[!duplicated_rows, ]
# keep only non-duplicated patientID --> dim(phenotype_df.new)=c(48,3)
df_ratios <- merge(df_ratios, phenotype_df, by.x = "patient", by.y = "patient")
df_ratios <- df_ratios[order(df_ratios[,phenotype_name]), ]
df_ratios$meta <- factor(df_ratios$meta, levels = unique(df_ratios$meta))


# Plot the stacked bar plot
ggplot(df_ratios, aes(x = meta, y = ratio, fill = PENCIL_pred_label)) +
  geom_bar(stat = "identity") +
  labs(x = "Patient", y = "Count ratio", fill = "Response") +
  theme(
    panel.background = element_rect(fill = "white"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) # +
  #scale_fill_manual(values = c("yes" = "#A6CEE3FF", "no" = "#5AA3DAFF"))

ggsave(filename = paste0(fig_dir, "stackedBar_cellRatio_", cellType_of_interest, "_",sampleType_name,".png"), height = 120, width = 190*1.2, dpi = 800, units = "mm")

```

# Calculate known ICB signature UCell score 
```{r}
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("UCell")
library(UCell)

ICB_sig <- list(
    exh_sig = c('PDCD1', 'CTLA4', 'LAG3', 'HAVCR2', 'TIGIT'), # blood scRNAseq HNSCC ms
    cyt_sig = c("GZMA", "PRF1"), # blood scRNAseq HNSCC ms
    CXCL13_sig = c("CXCL13"), # blood scRNAseq HNSCC ms
    periT_sig = c('GPR18', 'HLA-DOA', 'STAT1'), # Y. Bareche et al. 2022 AoO
    STAT1_sig = c('STAT1', 'GBP1', 'IRF1', 'WARS', 'PLAAT4', 'CXCL9', 'GBP5', 'LAP3', 'APOL3', 'IDO1', 'GBP4', 'LAG3', 'CXCL10', 'GZMA', 'PRF1', 'IFNG', 'GBP2', 'CCR5', 'SLC31A2', 'APOL6', 'CXCL11', 'SERPINA1', 'FCGR1B', 'TNFSF10', 'GZMK', 'CST7', 'FGL2', 'TAP1', 'SERPING1', 'FCER1G', 'CD2', 'GIMAP4', 'IFIT3', 'CCL4', 'CXCR6', 'GIMAP5', 'XAF1', 'SLAMF7', 'ANKRD22'), # Y. Bareche et al. 2022 AoO
    TCI0_sig = c('CCL5', 'CD27', 'CD274', 'CD276', 'CD8A', 'CMKLR1', 'CXCL9', 'CXCR6', 'HLA-DQA1', 'HLA-DRB1', 'HLA-E', 'IDO1', 'LAG3', 'NKG7', 'PDCD1LG2 '), # Y. Bareche et al. 2022 AoO
    TIS_sig = c('CD276', 'HLA-DQA1', 'CD274', 'IDO1', 'HLA-DRB1', 'HLA-E', 'CMKLR1', 'PDCD1LG2', 'PSMB10', 'LAG3', 'CXCL9', 'STAT1', 'CD8A', 'CCL5', 'NKG7', 'TIGIT', 'CD27', 'CXCR6'), # Y. Bareche et al. 2022 AoO
    cyt2_sig = c('PRF1', 'GZMB', 'GZMA', 'GZMH', 'NKG7', 'GNLY'), # (Mathewson et al., 2021)
    Teff_sig = c('CD8A', 'EOMES', 'PRF1', 'IFNG', 'CD274'), # David F. McDermott et al. 2018 Nat Med
    Teff_IFNG_sig = c('CD8A', 'GZMA', 'GZMB', 'IFNG', 'EOMES', 'CXCL9', 'CXCL10', 'TBX21'), # Louis Fehrenbacher et al 2016 Lancet
    NKR_sig = c('KLRD1', 'FGFBP2', 'FCGR3A', 'S1PR5', 'KLRC1', 'KLRC3', 'KLRB1', 'KLRC2'), # Luoma et al. Cell 2022
    preIFNG_sig = c('IFNG', 'STAT1', 'CCR5', 'CXCL9', 'CXCL10', 'CXCL11', 'IDO1', 'PRF1', 'GZMA', 'HLA-DRA'), # Ayers et al. 2017 JCI
    preTCI_sig = c('IL2RG', 'CXCR6', 'CD3D', 'CD2', 'ITGAL', 'TAGAP', 'CIITA', 'HLA-DRA', 'PTPRC', 'CXCL9', 'CCL5', 'NKG7', 'GZMA', 'PRF1', 'CCR5', 'CD3E', 'GZMK', 'IFNG', 'HLA-E', 'GZMB', 'PDCD1', 'SLAMF6', 'CXCL13', 'CXCL10', 'IDO1', 'LAG3', 'STAT1', 'CXCL11'), # preliminary expanded immune sig. Ayers et al. 2017 JCI
    IFNG_sig = c('IDO1', 'CXCL10', 'CXCL9', 'HLA-DRA', 'IFNG', 'STAT1'), # Ayers et al. 2017 JCI
    TCI_sig = c('CD3D', 'IDO1', 'CIITA', 'CD3E', 'CCL5', 'GZMK', 'CD2', 'HLA-DRA', 'CXCL13', 'IL2RG', 'NKG7', 'HLA-E', 'CXCR6', 'LAG3', 'TAGAP', 'CXCL10', 'STAT1', 'GZMB'), # Ayers et al. 2017 JCI,
    prolif_sig = c('BUB1', 'CCNB2', 'CDK1', 'CDKN3', 'FOXM1', 'PCLAF', 'MAD2L1', 'MELK', 'MKI67', 'TOP2A'), # Pabla et al. 2019 JITC
    TRM_sig1 = c("IL17A", "IL10", "IL2", "IFNG", "DUSP6", "PDCD1", "CRTAM", "ITGA1", "ITGAE", "CD69", "CD101", "CXCR6", "SELL-", "S1PR1-", "S1PR5-", "CX3CR1-", "MKI67-", "KLF2-", "KLF3-"), # "IL-17", "IL-10", "IL-2", "IFN-r", "DUSP6", "PD-1", "CRTAM", "CD49a", "CD103", "CD69", "CD101", "CXCR6", "CD62L-", "S1PR1-", "S1PR5-", "CX3CR1-", "Ki67-", "KLF2-", "KLF3-": Kumar et al. 2017 Cell Reports
    TRM_sig2 = c("ITGAE","VIM","ITGA1","TNFSF9") # Poon et al. Nature Immunology 2023
)

if('exh_sig' %in% colnames(merged_seu@meta.data)) {
  # merged_seu@meta.data$CD4T = NULL
  1
} else {
  merged_seu <- AddModuleScore_UCell(merged_seu, features=ICB_sig, name=NULL)
}


```

# Violin plot of know gene scores for N/NR cells
```{r}

feats <- c("exh_sig", "cyt_sig", "CXCL13_sig", "periT_sig", "STAT1_sig", "TCI0_sig", "TIS_sig", "cyt2_sig", "Teff_sig", "Teff_IFNG_sig", "NKR_sig", "preIFNG_sig", "preTCI_sig", "IFNG_sig", "TCI_sig", "prolif_sig", "TRM_sig1", "TRM_sig2") 
p <- VlnPlot(merged_seu, group.by = "PENCIL_pred_label", features = feats, pt.size = 0, ncol = 6) +
  NoLegend()
ggsave(plot = p,path = fig_dir,  filename = paste0("Vln_",cellType_of_interest,"_ICBsig_PENCILpredLabel_",sampleType_name,".pdf"),  width = 14, height = 8)

```


# Known ICG differential expressions (79 ICGs) with FDR < 0.05 in R/NR cells
```{r}

ICG_list = c('ADORA2A', 'BTLA', 'BTN2A1', 'BTN2A2', 'BTN3A1', 'BTNL3', 'BTNL9', 'C10orf54', 'CD160', 'CD209', 'CD226', 'CD27', 'CD274', 'CD276', 'CD28', 'CD40', 'CD40LG', 'CD47', 'CD70', 'CD80', 'CD86', 'CD96', 'CEACAM1', 'CTLA4', 'HAVCR2', 'HLA-A', 'HLA-B', 'HLA-C', 'HLA-DMA', 'HLA-DMB', 'HLA-DOA', 'HLA-DOB', 'HLA-DPA1', 'HLA-DPB1', 'HLA-DQA1', 'HLA-DQB1', 'HLA-DRA', 'HLA-DRB1', 'HLA-DRB3', 'HLA-DRB4', 'HLA-DRB5', 'HLA-E', 'HLA-F', 'HLA-G', 'ICOS', 'ICOSLG', 'IDO1', 'KIR2DL1', 'KIR2DL2', 'KIR2DL3', 'KIR2DL4', 'KIR2DL5A', 'KIR2DL5B', 'KIR2DS1', 'KIR2DS2', 'KIR2DS3', 'KIR2DS4', 'KIR2DS5', 'KIR3DL1', 'KIR3DL2', 'KIR3DL3', 'KIR3DS1', 'LAG3', 'LGALS9', 'PDCD1', 'PDCD1LG2', 'PVR', 'SIRPA', 'TDO2', 'TIGIT', 'TNFRSF14', 'TNFRSF18', 'TNFRSF4', 'TNFRSF9', 'TNFSF14', 'TNFSF18', 'TNFSF4', 'TNFSF9', 'VTCN1')

p <- VlnPlot(merged_seu, group.by = "PENCIL_pred_label", features = ICG_list[1:40], pt.size = 0, ncol = 10) +
  NoLegend()+
  labs(x = "", y ="")
ggsave(plot = p,path = fig_dir,  filename = paste0("Vln_",cellType_of_interest,"_ICG1_PENCILpredLabel_",sampleType_name,".pdf"),  width = 14, height = 8)

p <- VlnPlot(merged_seu, group.by = "PENCIL_pred_label", features = ICG_list[41:79], pt.size = 0, ncol = 10) +
  NoLegend()+
  xlab("") +
  ylab("") 
ggsave(plot = p,path = fig_dir,  filename = paste0("Vln_",cellType_of_interest,"_ICG2_PENCILpredLabel_",sampleType_name,".pdf"),  width = 14, height = 8)

```


# Identify and plot marker genes for N/NR cells
```{r}

logFCfilter = 0.25 # 0.5
adjPvalFilter = 0.05

##### find all markers for PENCIL predicted labels
Idents(object = merged_seu) <- merged_seu@meta.data$'PENCIL_pred_label'
Cells.markers = FindAllMarkers(merged_seu, only.pos=T, min.pct=0.25, logfc.threshold = logFCfilter)
Idents(object = merged_seu) <- merged_seu@meta.data$'seurat_clusters'

sig.markers = Cells.markers[(abs(as.numeric(as.vector(Cells.markers$avg_log2FC))) > logFCfilter & 
                               as.numeric(as.vector(Cells.markers$p_val_adj)) < adjPvalFilter), ]

write.csv(sig.markers, paste0(processed_dir, "sig.markers_PENCILpredLabel_",cellType_of_interest,'_',sampleType_name,".csv"))

##### Keep only the top 10 genes for each PENCIL predicted label, ranked by logFC
topN = 40
Top10.markers <- sig.markers %>%
    group_by(cluster) %>%
    top_n(topN, avg_log2FC)

DoHeatmap(merged_seu, features = Top10.markers$gene, group.by = "PENCIL_pred_label", slot = "data") + NoLegend()
ggsave(filename = paste0(fig_dir, "Heatmap_",cellType_of_interest,'_Top10markers_',sampleType_name,".png"), height = 14, width = 14)

p <- DotPlot(merged_seu, 
             features = unique(Top10.markers$gene), 
             group.by = "PENCIL_pred_label",
             cols = c("lightgrey", "red")) + 
  labs(x="", y="")+
  #coord_flip()+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
ggsave(filename = paste0(fig_dir, "dotplot_",cellType_of_interest,'_Top10markers_',sampleType_name,".png"), plot = p, height = 160, width = 280, dpi = 800, units = "mm")


```


# Violin plot of marker gene scores for N/NR cells
```{r}

marker_sig = list(
  R_cell_marker = Top10.markers$gene[1:topN],
  Rej_cell_marker = Top10.markers$gene[(topN+1):(2*topN)],
  NR_cell_marker = Top10.markers$gene[(2*topN+1):(3*topN)],
  R_minus_NR_cell_marker = c(Top10.markers$gene[1:topN], paste0(Top10.markers$gene[(2*topN+1):(3*topN)],'-'))
)

if('NR_cell_marker' %in% colnames(merged_seu@meta.data)) {
  # merged_seu@meta.data$NR_cell_marker = NULL
  1
} else {
  merged_seu <- AddModuleScore_UCell(merged_seu, features=marker_sig, name=NULL)
}

# merged_seu$NRR <- merged_seu$R_cell_marker
# merged_seu$R_cell_marker <- merged_seu$NR_cell_marker
# merged_seu$NR_cell_marker <- merged_seu$NRR
# merged_seu$NRR = NULL

feats <- c("NR_cell_marker", "Rej_cell_marker", "R_cell_marker", "R_minus_NR_cell_marker")
p <- VlnPlot(merged_seu, group.by = "PENCIL_pred_label", features = feats, pt.size = 0, ncol = 3) +
  NoLegend()
ggsave(plot = p,path = fig_dir,  filename = paste0("Vln_",cellType_of_interest,"_Top10.marker_PENCILpredLabel_",sampleType_name,".pdf"),  width = 8, height = 8)

```

# Use marker gene signature score to predict ICB response in 0.1) the same training data; 0.2) the same training data using blood-derived score; 1) bulk RNAseq; 2) other cancer types.
```{r}

merged_seu$meta <- factor(merged_seu$meta, levels = unique(df_ratios$meta0))

feats <- c("NR_cell_marker", "Rej_cell_marker", "R_cell_marker", "R_minus_NR_cell_marker")
p <- VlnPlot(merged_seu, group.by = "meta", features = feats, pt.size = 0, ncol = 1) +
  NoLegend()
ggsave(plot = p,path = fig_dir,  filename = paste0("Vln_",cellType_of_interest,"_Top10.marker_patient_",sampleType_name,".pdf"),  width = 14, height = 8)


#saveRDS(object = merged_seu, file = paste0(data_dir, "seu_",sampleType_name,"_", cellType_of_interest, ".rds")) 

```

